exchanges/kraken/kraken_websocket.go

```diff
@@ -1105,211 +924,365 @@ func trim(s string) string {
 }
 
 // wsProcessCandles converts candle data and sends it to the data handler
-func (k *Kraken) wsProcessCandles(channelData *WebsocketChannelData, data []interface{}) error {
-	startTime, err := strconv.ParseFloat(data[0].(string), 64)
-	if err != nil {
-		return err
+func (k *Kraken) wsProcessCandles(channelName string, response []any, pair currency.Pair) error {
+	// 8 string quoted floats followed by 1 integer for trade count
+	dataStr, ok := response[1].([]any)
+	if !ok || len(dataStr) != 9 {
+		return errors.New("received invalid candle data")
+	}
+	data := make([]float64, 8)
+	for i := 0; i < 8; i++ {
+		s, ok := dataStr[i].(string)
+		if !ok {
+			return fmt.Errorf("received invalid candle data: %w", common.GetTypeAssertError("string", dataStr[i], "candle-data"))
+		}
+
+		f, err := strconv.ParseFloat(s, 64)
+		if err != nil {
+			return fmt.Errorf("received invalid candle data: %w", err)
+		}
+		data[i] = f
 	}
 
-	endTime, err := strconv.ParseFloat(data[1].(string), 64)
-	if err != nil {
-		return err
+	// Faster than getting it through the subscription
+	parts := strings.Split(channelName, "-")
+	if len(parts) != 2 {
+		return errBadChannelSuffix
 	}
+	interval := parts[1]
 
-	openPrice, err := strconv.ParseFloat(data[2].(string), 64)
-	if err != nil {
-		return err
+	k.Websocket.DataHandler <- stream.KlineData{
+		AssetType:  asset.Spot,
+		Pair:       pair,
+		Timestamp:  time.Now(),
+		Exchange:   k.Name,
+		StartTime:  convert.TimeFromUnixTimestampDecimal(data[0]),
+		CloseTime:  convert.TimeFromUnixTimestampDecimal(data[1]),
+		OpenPrice:  data[2],
+		HighPrice:  data[3],
+		LowPrice:   data[4],
+		ClosePrice: data[5],
+		Volume:     data[7],
+		Interval:   interval,
 	}
+	return nil
+}
 
-	highPrice, err := strconv.ParseFloat(data[3].(string), 64)
-	if err != nil {
-		return err
+// GetSubscriptionTemplate returns a subscription channel template
+func (k *Kraken) GetSubscriptionTemplate(_ *subscription.Subscription) (*template.Template, error) {
+	return template.New("master.tmpl").Funcs(template.FuncMap{"channelName": channelName}).Parse(subTplText)
+}
+
+func (k *Kraken) generateSubscriptions() (subscription.List, error) {
+	return k.Features.Subscriptions.ExpandTemplates(k)
+}
+
+// Subscribe adds a channel subscription to the websocket
+func (k *Kraken) Subscribe(in subscription.List) error {
+	in, errs := in.ExpandTemplates(k)
+
+	// Collect valid subs to subscribe to; Note that we won't RemoveSub on any that err on SetState or AddSub
```

Comment:
I'm not sure I understand this question fully.
Could you expand on both sentences please?

* I think you're asking why we don't RemoveSub.

We don't remove existing subscriptions because it would be wrong to get a "Duplicate subscription" error, and then actually *remove* the subscription you've found was duplicated.

* You said that the knowledge is helpful to prevent changes down the line.

Assuming "down the line" means later in this function ... the subs which did not get Added are not collected as valid subs ... so ... there's no changes going to happen to them.

I'm thinking "say less" would have been better here to avoid confusion.

Fixed fd5bc1cf

Would this help clarify?

Reply:
Much better 🎉  It seemed like you were stating that something exceptional was happening. Down the line just meant, someone looking at the code without the context of this PR in future and potentially change said exceptional circumstance for the worse. I'll try not to use that phrasing given that we care about code lines 😄